src/coreclr/vm/classlayoutinfo.cpp

```diff
@@ -282,16 +281,32 @@ namespace
 
             pManagedPlacementInfo->m_size = (pNestedType.GetMethodTable()->GetNumInstanceFieldBytes());
 
-            if (pNestedType.GetMethodTable()->HasLayout())
+#if !defined(TARGET_64BIT) && (DATA_ALIGNMENT > 4)
+            if (pManagedPlacementInfo->m_size >= DATA_ALIGNMENT)
             {
-                pManagedPlacementInfo->m_alignment = pNestedType.GetMethodTable()->GetLayoutInfo()->m_ManagedLargestAlignmentRequirementOfAllMembers;
+                pManagedPlacementInfo->m_alignment = DATA_ALIGNMENT;
             }
             else
+#elif defined(FEATURE_64BIT_ALIGNMENT)
+            if (pNestedType->RequiresAlign8())
             {
+                pManagedPlacementInfo->m_alignment = 8;
+            }
+            else
+#endif // FEATURE_64BIT_ALIGNMENT
+            if (pNestedType.GetMethodTable()->ContainsPointers())
+            {
+                // this field type has GC pointers in it, which need to be pointer-size aligned
+                // so do this if it has not been done already
                 pManagedPlacementInfo->m_alignment = TARGET_POINTER_SIZE;
```

Comment:
What about types that contain both GC pointers and 64-bit fields in <code>RequiresAlign8</code> mode (on arm32), won't this override the previously set 8-alignment?

Reply:
There’s an `else` above in the Align8 case so this won’t be hit then.

Reply:
Right you are, sorry for overlooking, this looks good to me now.